pkg/webhook/certs.go

```diff
@@ -0,0 +1,486 @@
+
+var _ handler.Mapper = &mapper{}
+
+type mapper struct {}
+
+func (m *mapper) Map(object handler.MapObject) []reconcile.Request {
+	if object.Meta.GetNamespace() != vwhKey.Namespace {
+		return nil
+	}
+	if object.Meta.GetName() != vwhKey.Name {
+		return nil
+	}
+	return []reconcile.Request{{NamespacedName: secretKey}}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func addController(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New(("validating-webhook-controller"), mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to the provided ValidatingWebhookConfiguration
+	s := &corev1.Secret{}
+	if err := c.Watch(&source.Kind{Type: s}, &handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+
+	vwh := &unstructured.Unstructured{}
+	vwh.SetGroupVersionKind(vwhGVK)
+	mapper := &handler.EnqueueRequestsFromMapFunc{ToRequests: &mapper{}}
+	if err := c.Watch(&source.Kind{Type: vwh}, mapper); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileVWH{}
+
+// ReconcileVWH reconciles a validatingwebhookconfiguration, making sure it
+// has the appropriate CA cert
+type ReconcileVWH struct {
+	client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a validatingwebhookconfiguration
+// object and makes sure the most recent CA cert is included
+func (r *ReconcileVWH) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	if request.NamespacedName != secretKey {
+		return reconcile.Result{}, nil
+	}
+	secret := &corev1.Secret{}
+	if err := r.Get(context.TODO(), request.NamespacedName, secret); err != nil {
+		if k8sErrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	vwh := &unstructured.Unstructured{}
+	vwh.SetGroupVersionKind(vwhGVK)
+	if err := r.Get(context.TODO(), vwhKey, vwh); err != nil {
+		if k8sErrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	if secret.GetDeletionTimestamp().IsZero() {
+		artifacts, err := buildArtifactsFromSecret(secret)
+		if err != nil {
+			log.Error(err, "secret is not well-formed, cannot update ValidatingWebhookConfiguration")
+			return reconcile.Result{}, nil
+		}
+		log.Info("ensuring CA cert on ValidatingWebhookConfiguration")
+		injectCertToWebhook(vwh, artifacts.CertPEM)
+		if err := r.Update(context.Background(), vwh); err != nil {
+			return reconcile.Result{Requeue: true}, nil
```

Comment:
Why swallow the error? You could return it or at least log it?

Reply:
fixed